src/Security/AuthenticationMiddleware.php

```diff
@@ -60,4 +135,136 @@ public function process(HTTPRequest $request, callable $delegate)
+                function ($url) {
+                    return Director::absoluteURL($url);
+                },
+                $whitelist
+            ));
+
+            if (!in_array($currentUrl, $allowedUrls, true)) {
+                return $this->redirectOrForbid($request, $redirectUrl);
+            }
+        } else {
+            Security::setCurrentUser(null);
+        }
+    }
+
+    /**
+     * Returns a redirect to the URL if text/html is acceptable, otherwise
+     * deauthenticates the current request by Security::setCurrentUser(null)
+     *
+     * @param HTTPRequest $request
+     * @param string $redirectUrl
+     *
+     * @return HTTPResponse|null
+     */
+    private function redirectOrForbid(HTTPRequest $request, $redirectUrl)
+    {
+        $acceptableTypes = $request->getAcceptMimetypes();
+
+        $allowedTypes = [
+            '*/*',
+            'text/*',
+            'text/html',
+            'text/html',
+            'application/xhtml+xml',
+            'text/xml',
+            'application/xml'
+        ];
+
+        if (count(array_intersect($allowedTypes, $acceptableTypes)) > 0) {
+            $redirectAllowed = true;
+        } else {
+            //  if browser didn't send the Accept header
+            //  with mimetypes, let's redirect anyway
+            $redirectAllowed = count($acceptableTypes) === 0;
+        }
+
+        if ($redirectAllowed) {
+            $response = new HTTPResponse();
+            $response->redirect($redirectUrl);
+            return $response;
+        } else {
+            Security::setCurrentUser(null);
+        }
+    }
+
+    /**
+     * Preserve the password change URL in the session
+     * That URL is to be redirected to to force users change expired passwords
+     *
+     * @param Session $session Session where we persist the redirect URL
+     * @param string $url change password form address
+     */
+    public static function setExpiredPasswordRedirect(Session $session, $url)
+    {
+        $session->set(static::SESSION_KEY_EXPIRED_PASSWORD_REDIRECT, $url);
+    }
+
+    /**
+     * Allow a URL to be visited with expired password
+     *
+     * @param Session $session Session where we persist the whitelist
+     * @param string $url URL to be whitelisted
+     */
+    public static function whitelistExpiredPasswordUrl(Session $session, $url)
+    {
+        $whitelist = $session->get(static::SESSION_KEY_EXPIRED_PASSWORD_WHITELIST);
+        if (!is_array($whitelist)) {
+            $whitelist = [];
+        }
+        $whitelist[] = $url;
+        $session->set(static::SESSION_KEY_EXPIRED_PASSWORD_WHITELIST, $whitelist);
+    }
+
+    /**
+     * Allow the current request to be finished without password expiration check
+     *
+     * @param Session $session Session where we persist the redirect URL
+     * @param string $url change password form address
+     */
+    public static function expiredPasswordAllowCurrentRequest(Session $session)
```

Comment:
For "current request" stuff I feel as though it's cleaner to store detail on the request object rather than in session - if possible. With PSR-7 you have "attributes" on a server request for detail like this.

Reply:
with `PSR-7` yes, but where would you put it in our `HTTPRequest` implementation?

Reply:
There's a `__set`/`__get` API that sets "params" on the request. That's probably closest?

Reply:
Or an `X-` header? Or just the `$request->offsetSet($var, $val)`.

As long as it's properly namespaced, I think any of those will do.

Reply:
> There's a __set/__get API that sets "params" on the request. That's probably closest?

Well, that wouldn't be `PSR-7` compatible since "attributes" should be its own property on the class. Assigning through `__set` feels _dirty_ as it's same as assigning to non-existing public attribute of the Request instance... Not sure it's much better than keeping it in session? Furthermore, I'm not exactly sure we have guarantees that we are working with the same exact instance of Request everywhere across request handlers and middlewares. Without a proper PSR-7 `attributes` implementation that's risky to rely on that.

> Or an X- header? Or just the $request->offsetSet($var, $val)

That would require introducing some encryption, otherwise those could be forged and a malicious user would be able to avoid the change password requirement completely. It feels like such effort might not really worth the benefit?